### ISSUE TITLE
kanidm_unixd: Support PAM authentication via FIDO2/WebAuthn/"Passkeys"

### ISSUE THREAD

**Issue by JJGadgets** [kanidm/kanidm#1553]
**Is your feature request related to a problem? Please describe.**

Currently, my understanding is that `kanidm_unixd` only supports passwords for PAM authentication, and not FIDO2/WebAuthn/"Passkeys".

**Describe the solution you'd like**

FIDO2 as PAM authentication is shown to be possible as shown by pam_u2f.so (https://github.com/Yubico/pam-u2f), and the FIDO2 enrollment simply stores the FIDO2 challenge data to ~/.config/Yubico/u2f_keys or user-defined location of authfile.

I think it would be nice if `kanidm_unixd` could also support FIDO2 PAM authentication in its own PAM module, and if a user has both supported passkeys (e.g. YubiKeys and not iOS-only) and password enabled, allow a PAM variable to either prompt or choose which is preferred.

**Describe alternatives you've considered**

Alternatively, `kanidm_unixd` could "sync" or at least provide the FIDO2 challenge data stored within Kanidm's database to the enrolled Linux machine running pam_u2f.so so the authfile can be pointed to the `kanidm_unixd` "synced" file.

**Additional context**

Currently, all my personal Linux workstations (not my personal servers) use YubiKeys in passwordless FIDO2 mode for PAM authentication including PIN verification. Since Kanidm is embracing passwordless FIDO2 as a valid and preferred authentication method, I think this feature would be nice to round out the "passwordless-FIDO2-first" authentication provider experience.

**Comment by Firstyear** [comment]
I agree that this would be a good enhancement - I've been recently thinking about it too. So far we haven't had many other users with linux desktop machines which is probably why this hadn't been done sooner.

The main things would be:

* The pam module has to detect correctly if the user is at a local or remote session type to offer the correct auth methods - you can't do webauthn over ssh!
* We need to have a way to mark passkeys/credentials as unix auth capable - this way their public keys are added to the unix user token record for caching on the machine (for when you are out and about). As you correctly identify, iOS credentials can't be used for this without something like caBLE support.
* We need to probably alter the pam to unixd communication format to allow the pam module itself to interact with the usb device, since the unixd daemon is isolated away from the seat/session. That means we'd need some extra parts to correctly do this since the current pam code is assuming a certain workflow.

**Comment by leoleoasd** [comment]
And this is another usecase of #1703: We can do webauthn on remote machines via one-time-link.

**Comment by Firstyear** [comment]
> And this is another usecase of #1703: We can do webauthn on remote machines via one-time-link.

Not really. This is for CTAP2 on a local machine where you might need offline authentication. Imagine a laptop with no internet and you want to auth with a security key.

Anyway, some of the values here may be helpful to determine local vs remote session setup, which is critical for us to resolve this issue. 

https://github.com/systemd/systemd/blob/main/src/login/pam_systemd.c#L853

**Comment by kenjitamura** [comment]
> Describe alternatives you've considered
> 
> Alternatively, kanidm_unixd could "sync" or at least provide the FIDO2 challenge data stored within Kanidm's database to the enrolled Linux machine running pam_u2f.so so the authfile can be pointed to the kanidm_unixd "synced" file.

I see that sssd implemented passkey support earlier this year.  If the second option is considered could an implementation like theirs be used where the mapping data is exposed as an ldap attribute?  

https://sssd.io/release-notes/sssd-2.9.0.html
https://sssd.io/release-notes/sssd-2.9.2.html
![Screenshot from 2023-11-05 14-01-07](https://github.com/kanidm/kanidm/assets/5234739/558779dd-e810-484b-908b-22c01c004e49)

**Comment by Firstyear** [comment]
IIRC it's not actually doing passkeys/CTAP2 locally, it pops a qr code to scan to go to an oauth2 portal. It's passkey storage also is extremely primitive compared to ours. 

I'd rather not imitate anything about how SSSD/IPA do this, because they are significantly behind us in their passkey support.

**Comment by Firstyear** [comment]
@kenjitamura Also another issue here is we don't have a way in pam via something like sudo to distinguish local user vs remote an ssh, so we don't actually know properly if we can proceed with ctap2 or not. We need to actually have that solved for a good user experience.

**Comment by kenjitamura** [comment]
Thank you for the response! 

I think since they've pulled in libfido2 as a dependency they've offloaded the CTAP2 handling to it or at least in the blog post write up I found there isn't mention of the need for an oauth2 portal during the setup.  But maybe they just left it off in their steps:  
https://fedoraproject.org/wiki/Changes/Passkey_authentication_centrally_managed_users#How_To_Test

It looks like for the sudo problem, with FreeIPA specifically, they had this to say: 

> FreeIPA extension to issue Kerberos tickets based on the passkey authentication allows to solve usability issues in accessing network resources in a passwordless way. This extension also provides Kerberos authentication indicator support, making passkey authentication visible to Kerberos services. This can be used, for example, for passwordless SUDO access with pam_sss_gss module when a Kerberos ticket was obtained with a specific (passkey) authentication mechanism. 

Which I guess just sounds like if the authentication succeeds with a passkey it's automatically backed by kerberos and pulls in the [module they have to facilitate passwordless sudo](https://www.mankier.com/8/pam_sss_gss)?

Unfortunately a lot of this is over my head and it's probably something you've already seen.

https://ikerexxe.github.io/idm/2022/12/19/passkey-central-auth.html
https://fedoraproject.org/wiki/Changes/Passkey_authentication_centrally_managed_users#Benefit_to_Fedora

**Comment by yaleman** [comment]
> Thank you for the response!
> 
> I think since they've pulled in libfido2 as a dependency they've offloaded the CTAP2 handling to it or at least in the blog post write up I found there isn't mention of the need for an oauth2 portal during the setup. But maybe they just left it off in their steps: https://fedoraproject.org/wiki/Changes/Passkey_authentication_centrally_managed_users#How_To_Test

This describes using a locally connected Passkey (as in a physical token), which won't work for SSH-connected users - which is central to our problem - PAM can't tell the difference between a local user and an SSH-connected so you'd end up locking users out of remote authentication.

**Comment by MrFastDie** [comment]
Any updates on this?

**Comment by Firstyear** [comment]
We have been working to refactor and improve the pam/nss code to be able to support this in future. The primary issue at the moment is that linux's auth infrastructure doesn't easily allow distinction of a local/remote auth attempt.

**Comment by m00nwtchr** [comment]
You could add the feature locked behind a flag to the pam module, so it can be manually configured to be only used with local applications, like screen lockers etc.

**Comment by Firstyear** [comment]
We don't want to stick this behind featuire flags, because that doesn't solve the problem, it's a bandaid. Imagine you unable fido2 on your laptop, but then you need to ssh into your laptop to restore it for some reason (bad update etc). Suddenly you can't because you enable the fido2 support. 

We want to do this right, and that means knowing what services are or are not local, and we have to find a way to do that correctly.

**Comment by m00nwtchr** [comment]
No, I mean PAM itself allows you to only add a certain module to the "stack" of each app with separate files under `/etc/pam.d`, and for each app you can pass different options to each loaded PAM module. You could add a PAM configuration option to the `pam_kanidm.so` module which allows the user to choose whether that app will require FIDO2 authentication or not.

**Comment by Firstyear** [comment]
Almost every distro uses a pattern where they include a common-auth configuration, so this doesn't solve the issue, since pam_kanidm.so is only in that common file, not the per service file.

That's why we considered something like detecting the calling service during the pam interaction (eg sshd or sudo or gdm) and using that calling service "type" to determine if it's local console on not.

The issue though is sudo/su. Because by that point you're already authenticated and then we don't have a way to know if you are physically at the machine or not. So you would unlock your screen with fido2 but we'd have to force pw-only for sudo/su because we can't tell if you're at the machine, or if you're sshd in. 

That's the problem here, we need a way to determine the locality of the session for subsequent pam calls like sudo/su. If you have any ideas on how to solve that, that's the barrier to this proceeding. Without that, we would need to do the above where you just use fido on a graphical login/screen unlock only.

**Comment by m00nwtchr** [comment]
Maybe some of this could work? https://unix.stackexchange.com/questions/9605/how-can-i-detect-if-the-shell-is-controlled-from-ssh/

IIRC PAM modules are loaded into the process doing the authentication, e.g. sudo, so the environment variables should be present. The `pam_lastlog` module seems to be able to determine some information about user sessions.

**Comment by Firstyear** [comment]
That seems like a possible avenue to explore and would potentially solve the issues here.

**Comment by Gabgobie** [comment]
Hi,

a few days ago I read this issue and something felt off to me.
I'd like to re-ignite discussion here and maybe outline a few options to progress.

First, I think we need to clearly define why and in which situations we would want a desktop login via Kani in the first place.

I currently can only think of two situations like that, one of which is not yet applicable.

1. As an administrator of a fleet of devices, I'd like any employee to be able to securely (not with shared accounts/passwords) login to their own account on any machine in the org (or OU).
1. Depending on the outcome of #2627, desktop login may be necessary to access fileshares. I'll go out on a limb and say desktop login _really shouldn't_ be necessary for that. To provide an example of the expected behavior, my current personal device can authenticate to a fileshare without being fully enrolled in the Kerberos realm. A local user is configured with Kerberos credentials for authentication to remote services. No need for any IDM interaction for desktop login.

Since `kanidm_unixd`, to my understanding, does not provide any Kerberos alike SSO behavior (correct me if I'm wrong), there is little benefit in using it aside from shared devices.

With that established, we can look into the currently blocking issues, namely SSH.

This is where things started to feel off to me. (Attested) Passkeys are (rightfully) regarded as the most secure level of authentication in Kanidm (until Smartcards (#2417) make it in).

The user is strongly discouraged from setting a password at all.
There is a warning text: `If possible, passkeys should be used instead, as they are phishing and exploit resistant.`
The `Add Password` button is *the only* yellow button in the self-service "Credentials" page.[^1]

At the same time using a password to remotely login to your device and as suggested above even accessing sudo privileges is regarded as a necessary evil:

> The issue though is sudo/su. Because by that point you're already authenticated and then we don't have a way to know if you are physically at the machine or not. So you would unlock your screen with fido2 but we'd have to force pw-only for sudo/su because we can't tell if you're at the machine, or if you're sshd in.

A common configuration on my Linux VMs, Desktops and Laptops is disabling password authentication for SSH entirely.

The actual goal here seems to be lockout prevention:

> Imagine you unable fido2 on your laptop, but then you need to ssh into your laptop to restore it for some reason (bad update etc). Suddenly you can't because you enable the fido2 support.

Now assuming this goal I would argue that everything we need is in place. Kani can already manage a user's SSH public keys. Password login _should not_ be an option for remote users _at all_, especially considering that the UNIX password cannot be part of a typical password manager and MUST be entered by a human manually. This results in notoriously weak passwords. We all know ["correct horse battery staple"](https://m.xkcd.com/936/) but have you ever asked a non-tech/security person to use a strong password? I know I have tried for years. 🥲

If something more advanced is needed, a SSH CA can be a great way to go about that. Smallstep has been mentioned in #3513 which unfortunately went a bit off the rails. I could both see Kani becoming a CA or clearly declaring this as an out of scope feature that has already been solved by other tools. In either case, using certificate based SSH allows for fine grained access controls and unless native support is planned, `step-ca` is likely the way I'll go for SSH authentication in the coming months.

Both SSH keys and SSH Certificates would still work if the client we are trying to access is unable to reach the Kanidm instance.

## Other considerations

### Privilege escalation

In many issues I've read, #4091 being the most recent, re-authentication for privilege escalation has been a topic. For sudo/su I found this stackexchange thread from 2014 with some interesting points: <https://security.stackexchange.com/questions/72689/ssh-key-auth-but-still-need-password-for-sudo>

Considering

> 1. As an administrator of a fleet of devices, I'd like any employee to be able to securely (not with shared accounts/passwords) login to their own account on any machine in the org (or OU).

I find myself wondering if IDM provisioned users have any business using sudo/su at all but I'm sure there is some valid use-case.

This leaves us with a few options.

- Don't re-authenticate the user for sudo/su
  I don't really like this one as a privileged user may forget to lock their session when going for lunch (we are still dealing with humans here) and that would leave the desktop entirely unprotected.
- Force a password for re-authentication as suggested in https://github.com/kanidm/kanidm/issues/1553#issuecomment-2635434432
  Honestly not as bad as it seems - the desktop is already unlocked and the user has access to a password manager. A complex password can be copy-pasted for privilege escalation. This is IMO the only acceptable use-case for the UNIX password. SSH is TOFU and being able to use the password for login exposes unnecessary attack surface. If the pw is only valid for privilege escalation, the session should already be established as trustworthy.
- SSH CAs are not really affected by this.
  They can (to my understanding) embed the exact allowed commands in the issued certificate. [Source: Smallstep - RBAC](https://smallstep.com/product/ssh) A native implementation could re-authenticate for an even shorter lived certificate with elevated privileges - I don't know how one would go about this for external tools.
- Maybe a custom integration directly with sshd but that's outside my understanding and seems unnecessarily complex when everything we need is already in place. That might also put the offline capability in question.

### Detecting SSH

Adding to https://github.com/kanidm/kanidm/issues/1553#issuecomment-2637905990 :

There seem to be some more environment variables to check: <https://stackoverflow.com/a/2908151>

- `SSH_CONNECTION`
- `SSH_CLIENT`
- `REMOTEHOST`
- `DISPLAY`
- `SESSIONNAME`

Alternatively, you can potentially also check the stdin and stdout: <https://stackoverflow.com/questions/2908044/how-to-tell-if-running-in-a-linux-console-versus-an-ssh-session#comment2959219_2908128>

Best,
Gab

[^1]: There is also no way to add/require a TOTP for UNIX auth by default but this is besides the point. I assume this can be enabled by an administrator using the client tools.

**Comment by Firstyear** [comment]
>     2. Depending on the outcome of [Ability to authenticate with Kanidm to Samba (epic) #2627](https://github.com/kanidm/kanidm/issues/2627), desktop login may be necessary to access fileshares. I'll go out on a limb and say desktop login _really shouldn't_ be necessary for that. To provide an example of the expected behavior, my current personal device can authenticate to a fileshare without being fully enrolled in the Kerberos realm. A local user is configured with Kerberos credentials for authentication to remote services. No need for any IDM interaction for desktop login.
> 
> 
> Since `kanidm_unixd`, to my understanding, does not provide any Kerberos alike SSO behavior (correct me if I'm wrong), there is little benefit in using it aside from shared devices.

It is being worked on in the background. Samba client/server with krb is on the roadmap. 

> 
> A common configuration on my Linux VMs, Desktops and Laptops is disabling password authentication for SSH entirely.

Yes, disabling pw authentication *is* one option. You have to use an attested ssh key with a pin or something, and that gives you "MFA" to have even been able to remotely access the system. 

However, not everyone wants to disable sudo passwords, and there are of course good reasons for that. It's never an absolute with security, it's a spectrum. We have to be mindful in Kanidm that this spectrum exists, and how that impacts our designs.

I think if Kani were to say "sudo must not have a password" then a lot of people would be very unhappy with us. 

> 
> The actual goal here seems to be lockout prevention:
> 
> > Imagine you unable fido2 on your laptop, but then you need to ssh into your laptop to restore it for some reason (bad update etc). Suddenly you can't because you enable the fido2 support.

In a way it's lockout prevention, but it's more - linux login sessions are too dumb to tell you if you are remote or at the console. So imagine you ssh to a machine in order to do work, but then sudo doesn't work as you aren't physically at the console. That's the issue. 

> 
> Now assuming this goal I would argue that everything we need is in place. Kani can already manage a user's SSH public keys. Password login _should not_ be an option for remote users _at all_, especially considering that the UNIX password cannot be part of a typical password manager and MUST be entered by a human manually. This results in notoriously weak passwords. We all know ["correct horse battery staple"](https://m.xkcd.com/936/) but have you ever asked a non-tech/security person to use a strong password? I know I have tried for years. 🥲

However, part of the design of Kanidm is that we *KNOW* the unix password weaker because it's memorised as you say. That's why we *strictly limit where it is used and the surface area to attack it*. 

In theory this should make the unix password only needed for sudo, but we see lots of people using it for other things today (due to a number of factors).


> 
> If something more advanced is needed, a SSH CA can be a great way to go about that. Smallstep has been mentioned in [#3513](https://github.com/kanidm/kanidm/pull/3513) which unfortunately went a bit off the rails

Yes, the ticket creator didn't seem to understand what a CA was, and that we were not in fact, a CA. 

> . I could both see Kani becoming a CA or clearly declaring this as an out of scope feature that has already been solved by other tools. In either case, using certificate based SSH allows for fine grained access controls and unless native support is planned, `step-ca` is likely the way I'll go for SSH authentication in the coming months.
> 
> Both SSH keys and SSH Certificates would still work if the client we are trying to access is unable to reach the Kanidm instance.

Using ssh-ca or even kerberos would both resolve some of the issues re stronger attested login, meaning a pw isn't required for sudo. They both achieve the same thing (short term ssh credentials to a remote system). We have considered both paths, and haven't ruled either out yet. 

But neither of these actually resolve the issue in question in this ticket.

Even if you used ssh-ca or krb, people may still want a sudo password on remote systems. Just because you are comfortable with passwordless sudo, doesn't mean everyone is.

In our work, we have to account for both ends of this, and that means we have to be careful about how people access machines to prevent breakages/lockouts etc. 

> ## Other considerations
> ### Privilege escalation
> 
> In many issues I've read, [#4091](https://github.com/kanidm/kanidm/issues/4091) being the most recent, re-authentication for privilege escalation has been a topic. For sudo/su I found this stackexchange thread from 2014 with some interesting points: https://security.stackexchange.com/questions/72689/ssh-key-auth-but-still-need-password-for-sudo

Again, security is a spectrum. There are genuine reasons why someone may make one of the choices as shown in that thread. 

> 
> ### Detecting SSH
> 
> Adding to [#1553 (comment)](https://github.com/kanidm/kanidm/issues/1553#issuecomment-2637905990) :
> 
> There seem to be some more environment variables to check: https://stackoverflow.com/a/2908151
> 
>     * `SSH_CONNECTION`
> 
>     * `SSH_CLIENT`
> 
>     * `REMOTEHOST`
> 
>     * `DISPLAY`
> 
>     * `SESSIONNAME`

The question then becomes what if I do:

```
sudo -s
su myself
sudo -s
```

Or even

```
su someoneelse
sudo -s
```

The environment variables now don't pass through, and I now broke sudo password prompts.

So today the only way to truly detect this properly is to nominate which interfaces are guaranteed to be at a physical console (tty, gdm, whatever) and they can use ctap or whatever, and then everything else has to fall back to the potential worst case (remote etc). 

So I don't know how to solve this, maybe there is some way we can channel env vars through with our account/session modules, maybe there is a way we can track sessions or whatever, but we haven't had the time to investigate. 

We also haven't had large demands to investigate, most people use kanidm-unixd on servers (remote ssh only), and only a handful use it on laptops etc. We have had other priorities that are more important to us to pursue than this so far. 

> 
> 
> Alternatively, you can potentially also check the stdin and stdout: https://stackoverflow.com/questions/2908044/how-to-tell-if-running-in-a-linux-console-versus-an-ssh-session#comment2959219_2908128
> 
> Best, Gab
> ## Footnotes
> 
>     1. There is also no way to add/require a TOTP for UNIX auth by default but this is besides the point. I assume this can be enabled by an administrator using the client tools. [↩](#user-content-fnref-1-42050502be4de31b85a1b6b0887fdd44)

This is intentional, because then we don't have offline auth for roaming users as we have no way to distribute the TOTP seeds securely today.

**Comment by Gabgobie** [comment]
Thanks for the quick response.

> > Since kanidm_unixd, to my understanding, does not provide any Kerberos alike SSO behavior (correct me if I'm wrong), there is little benefit in using it aside from shared devices.
> 
> It is being worked on in the background. Samba client/server with krb is on the roadmap.

I've been loosely following along with the related issues. My point was about the kind of integration.

If I understand correctly, `kanidm_unixd` is going to be necessary to populate the ccache and access krb5 protected network shares (`SMB`/`NFS`). I think it should still be possible to use `kanidm_unixd` to access a network share when using a local user account and providing IDM credentials in some way but this is getting off-topic.

The behavior I was trying to illustrate is the difference in behavior.
When logging in to a kerberized host, it will handle all authentication events for me. Opening `Negotiate` protected website X and my desktop login already asserts my identity. No interaction from me.
I don't expect (although it would be nice) for a desktop login through `kanidm_unixd` to do the same. To my understanding I'd still have to authenticate to my Kani instance in the browser when accessing one of my services.

> However, not everyone wants to disable sudo passwords, and there are of course good reasons for that. It's never an absolute with security, it's a spectrum. We have to be mindful in Kanidm that this spectrum exists, and how that impacts our designs.

I'm not asking for sudo passwords to be disabled by force and was hoping for that to be clear later on in the post. I was doing the research while writing and the further I got the more certain I was that sudo passwords or the use of a ssh-ca are the best ways forwards.

> I think if Kani were to say "sudo must not have a password" then a lot of people would be very unhappy with us.

Me included! <3

> [...] linux login sessions are too dumb to tell you if you are remote or at the console. So imagine you ssh to a machine in order to do work, but then sudo doesn't work as you aren't physically at the console. That's the issue.

That much I got from the previous posts. But unless a good way around the issue pops up unannounced - which I doubt - we should figure out a solution that is better than weak passwords that potentially enable root access.

Hence my later suggestion to make re-authentication events a password only option, enabling initial-authentication events to be passwordless in the process.

> However, part of the design of Kanidm is that we KNOW the unix password weaker because it's memorised as you say. That's why we strictly limit where it is used and the surface area to attack it.
> 
> In theory this should make the unix password only needed for sudo, but we see lots of people using it for other things today (due to a number of factors).

This is a bit confusing to me. I'll try to split it up so you can catch where and if I misunderstand.

> That's why we strictly limit where it is used and the surface area to attack it.

Confirming: The UNIX password is separate from regular login and only valid for `kanidm_unixd` desktop login and re-auth operations.

> In theory this should make the unix password only needed for sudo, [...]

Is this following my suggestion of only using the UNIX *password* for re-auth and enabling hardware tokens for the initial desktop login or the current state? Because the way I understand the current state, the UNIX password is enabled for

1. Desktop login
1. SSH login
1. sudo/su privilege escalation

and thereby enables the weakest credential in the system to remotely access any device, escalate their privileges to root and thus install/modify any piece of software.

By my understanding, even if the password itself is as limited in scope as possible without compromising its justification for existing, it is among the credentials with the largest blast radius.

> [...] but we see lots of people using it for other things today (due to a number of factors).

Interesting. I'd like to read more about that and the rationale.

> Using ssh-ca or even kerberos would both resolve some of the issues re stronger attested login, meaning a pw isn't required for sudo. They both achieve the same thing (short term ssh credentials to a remote system). We have considered both paths, and haven't ruled either out yet.
> 
> But neither of these actually resolve the issue in question in this ticket.

It feels logical to tackle ssh if remote re-auth is the blocker for this ticket but maybe I misunderstand. Anyways, good to hear that both of the aforementioned paths are on the table.

> Even if you used ssh-ca or krb, people may still want a sudo password on remote systems. Just because you are comfortable with passwordless sudo, doesn't mean everyone is.

And people should be able to use sudo passwords! Unless I absolutely mangled my writing last night, that was the conclusion of my post.

To quote myself:

> > Force a password for re-authentication as suggested in https://github.com/kanidm/kanidm/issues/1553#issuecomment-2635434432
> > Honestly not as bad as it seems - the desktop is already unlocked and the user has access to a password manager. A complex password can be copy-pasted for privilege escalation. This is IMO the only acceptable use-case for the UNIX password. SSH is TOFU and being able to use the password for login exposes unnecessary attack surface. If the pw is only valid for privilege escalation, the session should already be established as trustworthy.

> In our work, we have to account for both ends of this, and that means we have to be careful about how people access machines to prevent breakages/lockouts etc.

The part I currently still take issue with is that, unless I _really_ misunderstood something, hardware tokens are ruled out for desktop login because they cannot be used over ssh and there is no way to distinguish local versus remote sessions so Kani doesn't know what mechanisms to offer for re-auth.

The current solution to this is to force passwords in all UNIX interactions instead of _enabling_ passkeys for initial login events such as unlocking the desktop.

The message I am trying to transmit is *not* to disable re-auth and passwords for all people. It is that needing a password for privilege escalation doesn't have to rule out a stronger login mechanism for initial authentication.

- I can login to a server using a SSH key and still need a password for sudo.
- I _should_ be able to login to a desktop using my hardware token and still need a password for sudo.
- Passwordless sudo is in the realm of user choice but not my cup of tea and not necessarily relevant to the topic aside from being a case where the blocker is a non-issue.

Forcing a password for re-auth is not an awful hack. Given I have to enter my password when using an ssh key unless I reconfigure my sudoers file it seems to be the standard.

By my account, `FIDO2 -> Password for escalaction` > `Password -> Same password for escalation`, which doesn't forbid users from setting their account up to be passwords only.

I'm tired and I'm starting to circle. My point should hopefully be sufficiently illustrated. Don't want to waste anyone's time talking besides each other.

Best,
Gab

Whoops, almost missed the last part!

> [...] and then everything else has to fall back to the potential worst case (remote etc).

Exactly my point. To avoid a fallback mechanism we currently live exclusively on the worst case. That feels backwards. Having to fall back to passwords is not a bad thing unless it exposes a downgrade attack. Remote login can be done with an ssh key, desktop unlock _can_ be done with a hardware token. To make it simple, privilege escalation would always be a password operation because at that point the user has already proven their identity with strong authentication and we only want to validate if it is still them on the other side or if it was intentional.

> [...] but we haven't had the time to investigate.

Absolutely fair point!

> [...] most people use kanidm-unixd on servers (remote ssh only), [...]

I'll assume they disabled passwords in sshd. Then `kanidm-unixd` provides pubkeys and makes them use passwords internally, for example for privilege escalation. I'd call that a display of the intended behavior. A strong mech for getting in and a password for privilege escalation. It would be great to have the same on desktop.

(I am also just realizing that I didn't think that far earlier. `kanidm-unixd` insisting on the use of passwords doesn't mean they are exposed to SSH)

> [...] and only a handful use it on laptops etc. [...]

I currently still don't really see the value in this. Desktops on the other hand... but I don't want to unnecessarily repeat myself ^^

> [...] We have had other priorities that are more important to us to pursue than this so far.

Again, fair!

**Comment by Firstyear** [comment]
> If I understand correctly, `kanidm_unixd` is going to be necessary to populate the ccache and access krb5 protected network shares (`SMB`/`NFS`). I think it should still be possible to use `kanidm_unixd` to access a network share when using a local user account and providing IDM credentials in some way but this is getting off-topic.

Sometimes those topics do align though - how we choose to let someone login has an influence on the credentials (such as krb tickets) that we issue. For example, if someone uses an ssh key, then how do we setup a ccache?

> That much I got from the previous posts. But unless a good way around the issue pops up unannounced - which I doubt - we should figure out a solution that is better than weak passwords that potentially enable root access.

Agreed. I have some ideas, but like I said, they are often blocked by if we are "local vs remote". I have some other ideas about using TPM PIN's which have hardware bruteforce protections. But nothing is "perfect".

> Hence my later suggestion to make re-authentication events a password only option, enabling initial-authentication events to be passwordless in the process.
>
> > However, part of the design of Kanidm is that we KNOW the unix password weaker because it's memorised as you say. That's why we strictly limit where it is used and the surface area to attack it.
> > In theory this should make the unix password only needed for sudo, but we see lots of people using it for other things today (due to a number of factors).
>
> This is a bit confusing to me. I'll try to split it up so you can catch where and if I misunderstand.
>
> > That's why we strictly limit where it is used and the surface area to attack it.
>
> Confirming: The UNIX password is separate from regular login and only valid for `kanidm_unixd` desktop login and re-auth operations.

Correct,

>
> > In theory this should make the unix password only needed for sudo, [...]
>
> Is this following my suggestion of only using the UNIX _password_ for re-auth and enabling hardware tokens for the initial desktop login or the current state? Because the way I understand the current state, the UNIX password is enabled for
>
>     1. Desktop login
>
>     2. SSH login
>
>     3. sudo/su privilege escalation
>
>
> and thereby enables the weakest credential in the system to remotely access any device, escalate their privileges to root and thus install/modify any piece of software.

That's correct, with a MAJOR caveat - we STRONGLY discourage the use of passwords directly on SSH. It's why we have a whole SSH public key distribution mechanism built in.

So you'd be correct, that anyone doing 2., it's the weakest link.

But if you have SSH key auth only, then we even [advise you disable password auth](https://kanidm.github.io/kanidm/stable/integrations/ssh_key_distribution.html#public-key-caching-configuration)

If you frame it *that* way, then you have:

1. SSH Key Only remote authentication
2. Desktop Login with Password

But of course, we can't control what people do. We *could* actually make Kanidm unixd reject password validation from the ssh service, which would harden against this, and probably is a good idea ....

> By my understanding, even if the password itself is as limited in scope as possible without compromising its justification for existing, it is among the credentials with the largest blast radius.

Sure, I agree. It just depends on the way you look at the radius. If someone gets the unix pw, they don't get the users whole account for example. That's one of the nice benefits of credential separation.

>
> > [...] but we see lots of people using it for other things today (due to a number of factors).
>
> Interesting. I'd like to read more about that and the rationale.

Generally it's IMAP/EMAIL because I haven't finished application passwords yet :(


> It feels logical to tackle ssh if remote re-auth is the blocker for this ticket but maybe I misunderstand. Anyways, good to hear that both of the aforementioned paths are on the table.

SSH is the blocker only because when you SSH in we don't know if you physically are at the console. If you SSH in and we try to prompt for CTAP, then you can't sudo at all, and you're screwed.


> The part I currently still take issue with is that, unless I _really_ misunderstood something, hardware tokens are ruled out for desktop login because they cannot be used over ssh and there is no way to distinguish local versus remote sessions so Kani doesn't know what mechanisms to offer for re-auth.

It's not that hardware tokens are ruled out for local auth at all - I'd love to make it happen!

But if we DO we want to make sure we do it RIGHT. And that means accounting for local and remote users!.

> The current solution to this is to force passwords in all UNIX interactions instead of _enabling_ passkeys for initial login events such as unlocking the desktop.
>
> The message I am trying to transmit is _not_ to disable re-auth and passwords for all people. It is that needing a password for privilege escalation doesn't have to rule out a stronger login mechanism for initial authentication.
>
>     * I can login to a server using a SSH key and still need a password for sudo.
>
>     * I _should_ be able to login to a desktop using my hardware token and still need a password for sudo.
>
>     * Passwordless sudo is in the realm of user choice but not my cup of tea and not necessarily relevant to the topic aside from being a case where the blocker is a non-issue.

And this is what we want to achieve too. I think we want the same thing! The issue as mentioned is we need to know about the *initial* authentication because it influences the *second*.

Alternately, we need to know if the authentication is local or remote, and that defines what we can offer.


> > [...] and then everything else has to fall back to the potential worst case (remote etc).
>
> Exactly my point. To avoid a fallback mechanism we currently live exclusively on the worst case. That feels backwards. Having to fall back to passwords is not a bad thing unless it exposes a downgrade attack. Remote login can be done with an ssh key, desktop unlock _can_ be done with a hardware token. To make it simple, privilege escalation would always be a password operation because at that point the user has already proven their identity with strong authentication and we only want to validate if it is still them on the other side or if it was intentional.

I agree. But linux auth is a mess, and it's the mess we have to live with. We are trying to work within the boundaries presented to us.

>
> > [...] but we haven't had the time to investigate.
>
> Absolutely fair point!
>
> > [...] most people use kanidm-unixd on servers (remote ssh only), [...]
>
> I'll assume they disabled passwords in sshd. Then `kanidm-unixd` provides pubkeys and makes them use passwords internally, for example for privilege escalation. I'd call that a display of the intended behavior. A strong mech for getting in and a password for privilege escalation. It would be great to have the same on desktop.
>
> (I am also just realizing that I didn't think that far earlier. `kanidm-unixd` insisting on the use of passwords doesn't mean they are exposed to SSH)
>
> > [...] and only a handful use it on laptops etc. [...]
>
> I currently still don't really see the value in this. Desktops on the other hand... but I don't want to unnecessarily repeat myself ^^

Well, what I'm saying more is very few people have the laptop or desktop use case. We're excited that people DO have this! But it's not been on the radar for a while as most people are server focused.

>
> > [...] We have had other priorities that are more important to us to pursue than this so far.
>
> Again, fair!

But also, don't think this means I am dismissing this. This feature is something I really want!

As mentioned, the issue is that if we do it, I want to do it right. That means we have to be careful about how and what we present, and ensure we don't lock someone out remotely, or prevent a desktop login.

So I think in the end, we "aggressively agree" if I am reading properly - we both want strong initial factors, and then sudo password is okay. The issue today is knowing if the initial factor is local or not. There are some ideas I have today, mostly about detecting if the calling pam service is one that is local or remote. It's probably the best path forward.

**Comment by Gabgobie** [comment]
> > If I understand correctly, kanidm_unixd is going to be necessary to populate the ccache and access krb5 protected network shares (SMB/NFS). I think it should still be possible to use kanidm_unixd to access a network share when using a local user account and providing IDM credentials in some way but this is getting off-topic.
> 
> Sometimes those topics do align though - how we choose to let someone login has an influence on the credentials (such as krb tickets) that we issue. For example, if someone uses an ssh key, then how do we setup a ccache?

Really interesting point. How _is_ a user authenticated with `kanidm-unixd`? I tried going through the docs and noticed that I assumed some kind of domain join, which is still in the making.

Going by <https://kanidm.github.io/kanidm/stable/integrations/ssh_key_distribution.html> I'd make the following assumptions:

The anonymous account is still a thing (although I thought I read somewhere that it was or is planned to be disabled) and ssh pubkeys are accessible to anyone, even without authentication.

- A user accessing a machine via SSH using any mechanism is authenticated through `PAM` with the [UsePam option](https://www.man7.org/linux/man-pages/man5/sshd_config.5.html#:~:text=UsePAM,-Enables). `"this will enable PAM authentication using KbdInteractiveAuthentication and PasswordAuthentication in addition to PAM account and session module processing for all authentication types"`
- `kanidm-unixd` is aware of everything (aside from the source of the request?)

or

- A user accessing a machine via SSH using an SSH key is authenticated by `sshd` using [AuthorizedKeysCommand](https://www.man7.org/linux/man-pages/man5/sshd_config.5.html#:~:text=AuthorizedKeysCommand) against `kanidm-unixd`.
- `kanidm-unixd` is aware of the attempted user authentication but not the status of the attempt

That leads me to believe that the second flow should be avoided. Is Kani missing the facilities to authenticate ssh key based login for the second flow to be necessary?

An offline authentication event against cached data would not result in a populated ccache. This could result in a bad UX when logging in offline and later wanting to access a network share unless there is a way to prompt for re-auth when the user is trying to access a service that requires an issued credential.
Online, the challenge would probably have to be forwarded to the Kani server for authentication and the ccache can be populated appropriately.

Since group membership etc. is to my understanding not anonymously accessible, I'm pretty sure that independently of the login mechanism, any desktop user that is logging in online must successfully authenticate against the Kani server - not unixd - so the necessary information for credential issuance should be available, no?

To really reason about this I'd probably need a better grasp on what information is available, what we do know and what we don't.

In terms of credential issuance through `kanidm-unixd` for local users, windows has the `credential manager` where I stored login data for an IPA domain at some point, Gnome has a section in the settings where you can enter login data for external accounts (again something I found when working with IPA) and I'd assume/hope that other environments have a similar keyring mechanism to connect local accounts to remote accounts without the need for desktop login with a non-local user. This use-case could _probably_ benefit from a dedicated credential.

> Agreed. I have some ideas, but like I said, they are often blocked by if we are "local vs remote". I have some other ideas about using TPM PIN's which have hardware bruteforce protections. But nothing is "perfect".

I'm not entirely sold on the entire TPM stuff because it is another hardware barrier and I generally like to avoid e-waste. Many of my devices don't have one. `tpm_if_possible` is very welcome though.

> We could actually make Kanidm unixd reject password validation from the ssh service, which would harden against this, and probably is a good idea ....

Would this go hand-in-hand with finding a solution for the issue of detecting the locality of the user or is detecting the requesting service already solved?

If that is already solved, a solution that by my understanding `does this right` would be to have a whitelist of sorts for services that can use CTAP such as `gdm` which are known to always be local, services that are always password only such as `sudo/su` and ssh, which could be locked to SSH keys (and maybe certificates depending on the amount of integration work or possibility for separation).

That would, if I am not entirely mistaken, not be a breaking change and satisfy this feature request to the greatest extent possible without sacrificing remote users.

Figuring out how to enable CTAP for privilege escalation of a local user could be considered an enhancement on top of this. The way OPs Workstations are setup suggests that they are already blocked from remote re-authentication events or have a fallback.

> The issue as mentioned is we need to know about the initial authentication because it influences the second.

Whereas my suggestion is to decouple the two if that actually helps. Initial auth *may* be strong. Second is locked to passwords.

> So I think in the end, we "aggressively agree" if I am reading properly

You do. Also, I adore that description and will yoink the term.

> There are some ideas I have today, mostly about detecting if the calling pam service is one that is local or remote. It's probably the best path forward.

Should have read to the end of your answer before searching the entirety of the internet for `PAM` and `sshd` behavior. We reached the same conclusion.

Dunno if it's worth mentioning but ssh is not necessarily remote (`ssh otheruser@localhost` instead of `su otheruser`). I'd still argue that once logged in, you should also be able to access a valid ssh key, for example resident to your hardware token or whichever secure distribution mechanism you choose so this _shouldn't_ be a lockout scenario.

**Comment by Firstyear** [comment]
>
> Really interesting point. How _is_ a user authenticated with `kanidm-unixd`? I tried going through the docs and noticed that I assumed some kind of domain join, which is still in the making.

You are "somewhat correct" but also "not" at the same time.

1. You SSH to the machine
1. The remote machine performs the authorised keys lookup via the command
1. kanidm-unixd contacts kanidmd either as anonymous or with a service-account token (domain join is later, that's a bit more involved).
1. kanidm-unixd retrieves the unix users information packet / token which contains the SSH public keys
1. kanidm-unixd returns the public keys to the authorised keys lookup
1. sshd validates the key the user presents as valid
1. sshd checks the users account validity via pam. This calls the pam-account stack
1. kanidm-unixd validates the account request, the groups of the user etc. This is the "authorisation step".
1. on success, sshd proceeds to create the users login session. This calls pam-session stack
1. kanidm-unixd is able to influence the state of the session.
1. sshd now hands off to the shell and the user is logged in.

> An offline authentication event against cached data would not result in a populated ccache.

Exactly. And that's why then when you go online you have to then populate the ccache in some manner.

But an offline auth event is the same as an ssh one - we don't have a user credential at any point, as sshd deals with the keys. How do we populate a ccache?

> > Agreed. I have some ideas, but like I said, they are often blocked by if we are "local vs remote". I have some other ideas about using TPM PIN's which have hardware bruteforce protections. But nothing is "perfect".
>
> I'm not entirely sold on the entire TPM stuff because it is another hardware barrier and I generally like to avoid e-waste. Many of my devices don't have one. `tpm_if_possible` is very welcome though.

TPM's will remain optional, they are a feature, not a requirement.

> > We could actually make Kanidm unixd reject password validation from the ssh service, which would harden against this, and probably is a good idea ....
>
> Would this go hand-in-hand with finding a solution for the issue of detecting the locality of the user or is detecting the requesting service already solved?

Semi related - we can "see" what pam service is calling into us, but we need to then label each service with if they are local or remote.

> If that is already solved, a solution that by my understanding `does this right` would be to have a whitelist of sorts for services that can use CTAP such as `gdm` which are known to always be local, services that are always password only such as `sudo/su` and ssh, which could be locked to SSH keys (and maybe certificates depending on the amount of integration work or possibility for separation).
>
> That would, if I am not entirely mistaken, not be a breaking change and satisfy this feature request to the greatest extent possible without sacrificing remote users.

Yes, this is one approach we think would work.

> Figuring out how to enable CTAP for privilege escalation of a local user could be considered an enhancement on top of this. The way OPs Workstations are setup suggests that they are already blocked from remote re-authentication events or have a fallback.
>
> > The issue as mentioned is we need to know about the initial authentication because it influences the second.
>
> Whereas my suggestion is to decouple the two if that actually helps. Initial auth _may_ be strong. Second is locked to passwords.

Yes, and that is also an option. Then sudo is always a password event, ssh can never be password, gdm is ctap etc.

> Dunno if it's worth mentioning but ssh is not necessarily remote (`ssh otheruser@localhost` instead of `su otheruser`). I'd still argue that once logged in, you should also be able to access a valid ssh key, for example resident to your hardware token or whichever secure distribution mechanism you choose so this _shouldn't_ be a lockout scenario.

This is also why we still have to assume the "worst cases" when ssh is involved. You may be at the console but WE DONT know that because it's via ssh. So we have to assume you are not able to touch the key.

**Comment by Gabgobie** [comment]
Thanks for the explanation of the flow!

> kanidm-unixd contacts kanidmd either as anonymous or with a service-account token

Is the use of service-account tokens documented or is that something that happens in the background? I didn't find anything about it in the [PAM and nsswitch section](https://kanidm.github.io/kanidm/stable/integrations/pam_and_nsswitch.html).

> we don't have a user credential at any point, as sshd deals with the keys. How do we populate a ccache?

I think there are two parts to this.

- The first one is that we need to get our hands on some kind of useful and strong credential to authenticate the user ourselves. Trusting `sshd` to do this for us is probably fine for gaining access to the system `sshd` is running on but certainly not for issuing credentials. Unfortunately, the few Ideas I had for this were all dead ends or hacky workarounds that I don't even want to suggest. In #2495 you mentioned that we could trust a domain-joined machine to authenticate users for us which I don't necessarily agree with. Consider the following scenario where an essentially untrusted host has to be domain joined:
  - Compliance/GDPR requirement to have no anonymous access
  - domain joined Docker host -> developers essentially get root access through the docker group
  - any person working with that host could now impersonate any other Kani user (from the `posix_group`?) because the "compromised" host is allowed to authenticate users
  - this could be used to gain access to another user's network share
  - I could have missed something here. I'm _really_ tired. Should this be raised in #2495?
- The second part is figuring out a way to issue credentials for offline users that come online and want to access a service. For now this is limited to populating the ccache. **If** we figure out a way to access the user's credential, this might be solvable with a refresh token. This token would be en- and decrypted with the user's secret. We'd also need to listen for attempts to use the gssapi so we know when to populate the ccache, retry until successful or find another meaningful trigger to retry getting the credentials. Maybe there's a better way.

The main blocker seems to again be ssh. To my understanding any mechanism other than ssh keys would be authenticated by `kanidm-unixd`. We could add a "privilege escalation"/"credential issue" command (UNIX password prompt) for non sudoers that want to trigger authentication through Kani. Not quite the desired SSO experience but on par with or better than now. Since ssh keys are already available for authentication, this issue will need to be solved for the `libkrimes` integration or whenever the ccache becomes a thing but it's probably out of scope for this issue.

The way I read the rest of your answer, we are in agreement about a possible path forwards. If it adds any value, I could try to formalize this in some way in the coming days.[^1]

[^1]: I'd have to take a look your other design documents to get a feel for how I would do this so it may very well take some time.

**Comment by Firstyear** [comment]
> Thanks for the explanation of the flow!
> 
> > kanidm-unixd contacts kanidmd either as anonymous or with a service-account token
> 
> Is the use of service-account tokens documented or is that something that happens in the background? I didn't find anything about it in the [PAM and nsswitch section](https://kanidm.github.io/kanidm/stable/integrations/pam_and_nsswitch.html).

I think it's documented in the next release. 

> 
> > we don't have a user credential at any point, as sshd deals with the keys. How do we populate a ccache?
> 
> I think there are two parts to this.

While I appreciate the thoughts here, those things are a separate topic unrelated to this one though :) 

> 
> The main blocker seems to again be ssh. To my understanding any mechanism other than ssh keys would be authenticated by `kanidm-unixd`. We could add a "privilege escalation"/"credential issue" command (UNIX password prompt) for non sudoers that want to trigger authentication through Kani. Not quite the desired SSO experience but on par with or better than now. Since ssh keys are already available for authentication, this issue will need to be solved for the `libkrimes` integration or whenever the ccache becomes a thing but it's probably out of scope for this issue.
> 
> The way I read the rest of your answer, we are in agreement about a possible path forwards. If it adds any value, I could try to formalize this in some way in the coming days.[1](#user-content-fn-1-cb9a2d3d1e2dfeb6dc10175fe2f39be0)
> ## Footnotes
> 
>     1. I'd have to take a look your other design documents to get a feel for how I would do this so it may very well take some time. [↩](#user-content-fnref-1-cb9a2d3d1e2dfeb6dc10175fe2f39be0)

I don't think we need anything too wild to move on.

We just need to add a list of pam service names that are considered "guaranteed local" which allow offering CTAP. And anything else falls back to PW or a remote viable authentication mechanism.

After that, it's just a method to distribute the public keys and mark them as login viable.

**Comment by Gabgobie** [comment]
> I don't think we need anything too wild to move on.

Great

> We just need to add a list of pam service names that are considered "guaranteed local" which allow offering CTAP. And anything else falls back to PW or a remote viable authentication mechanism.

I did some more digging. To my understanding X11 is designed as remote capable. If I am not mistaken (never used X11 remote), that means we can basically never, not even for display managers and desktop environments, claim that a service is "guaranteed local".

Looking for a list of services, I found <https://docs.oracle.com/cd/E19683-01/816-4883/pam-34/index.html> which directly contradicts the [sshd default](https://www.man7.org/linux/man-pages/man5/sshd_config.5.html#:~:text=moduli%2E-,PAMServiceName) of `sshd`.

I also went through my [`/etc/pam.d/<service_name>`](https://www.man7.org/linux/man-pages/man3/pam_start.3.html#:~:text=The%20policy%20for%20the%20service%20will%20be%20read%20from%20the%20file%20%2Fetc%2Fpam%2Ed%2Fservice%5Fname) which is likely not a comprehensive list to gather some more services.

Therefore I'd suggest adding an `assume_local: Vec<String>` (or `force_local`) key to the `unixd` configuration file. We would then offer CTAP for these services and provide a list of suggested services that _usually_ are local.

## `assume_local` suggestions

- sddm
- [gdm-password](https://unix.stackexchange.com/questions/473810/gnome-different-pam-configurations-for-lockscreen-vs-login#:~:text=GNOME%20uses%20gdm%2Dpassword%20as%20the%20PAM%20service%20name)
- kscreensaver
- sddm-greeter [(unsure)](https://manpages.debian.org/buster/sddm/sddm-greeter.1.en.html#:~:text=It%20is%20launched%20by%20the%20sddm%20daemon%2C%20end%20users%20are%20not%20meant%20to%20run%20it)
- kde
- kcheckpass
- ksu (unable to find info in reasonable amount of time but guess this is graphical privilege escalation)

<details><summary>Other <code>/etc/pam.d</code> filenames I found</summary>

## `treat_as_non_local` - service accounts, ssh keys, other mechanism, remote, unknown, no idea

### Display manager stuff

#### KDE/SDDM

- kde-fingerprint (I _think_ this is usually used against passkey-alike - sddm/sddm#284 sddm/sddm#1103)
- kde-smartcard
- sddm-autologin

#### GDM

- gdm-fingerprint
- gdm-autologin
- gdm-launch-environment
- gdm-smartcard

### Other

- login
- sshd
- su
- sudo
- sudo-i
- fingerprint-auth
- config-util
- vpnc-disconnect
- [postlogin](https://www.linux.org/docs/man5/postlogin.html)
- vpnc
- pluto [(seems to be used by `libreswan`)](https://github.com/libreswan/libreswan/wiki/HOWTO:-VPN-server-for-remote-clients-using-IKEv1-XAUTH-with-Certificates#userpassword-authentication-for-xauth)
- samba
- vmtoolsd
- chsh
- chfn
- password-auth
- sssd-shadowutils
- vlock
- runuser-l
- passwd
- [atd](https://linux.die.net/man/8/atd)
- anaconda
- runuser
- crond
- cups
- su-l
- cockpit
- remote
- switchable-auth
- smartcard-auth
- ppp
- other
- [system-auth](https://linux.die.net/man/5/system-auth)

</details>

I'm not very confident about many of them, especially without being about to test anything, so if anyone with some more PAM experience wants to chime in, that would be great.

**Comment by Firstyear** [comment]
X11 is dead anyone, so that's not a problem. 

Yes, that would be how I'd do it with a configuration option.
